test/replication-luatest/commit_wait_test.lua

```diff
@@ -1,58 +1,207 @@
+    cg.s:update_box_cfg{wal_queue_max_size = 1}
+end)
+
+g_corner_cases.test_full_journal_queue = function(cg)
+    cg.s:exec(function(wait_mode)
+        -- Fill up the journal queue.
+        box.atomic({wait = 'none'}, function() box.space.a:replace{0} end)
+
+        if wait_mode == 'none' then
+            local msg = 'The WAL queue is full'
+            t.assert_error_msg_content_equals(msg, function()
+                box.atomic({wait = 'none'}, function()
+                    box.space.s:replace{0}
+                end)
+            end)
+            return
         end
-    end)
+
+        local f = _G.fiber.create(function()
+            box.atomic({wait = wait_mode}, function()
+                box.space.s:replace{0}
+            end)
+        end)
+        f:set_joinable(true)
+        if wait_mode == 'submit' then
+            t.assert(f:join())
+        end
+        t.assert_equals(box.info.synchro.queue.len, 1)
+        box.atomic({txn_isolation = 'read-committed'}, function()
+            t.assert_equals(box.space.s:get{0}, {0})
+        end)
+        box.atomic({txn_isolation = 'read-confirmed'}, function()
+            t.assert_equals(box.space.s:get{0},
+                            _G.are_txns_not_isolated and {0} or nil)
+        end)
+
+        box.cfg{replication_synchro_quorum = 1}
+        t.helpers.retrying({timeout = 120}, function()
+            t.assert_equals(box.info.synchro.queue.len, 0)
+        end)
+        t.assert_equals(box.space.s:get{0}, {0})
+    end, {cg.params.wait_mode})
+end
+
+g_corner_cases.before_test('test_full_limbo', function(cg)
+    cg.s:update_box_cfg{replication_synchro_queue_max_size = 1}
+end)
+
+g_corner_cases.test_full_limbo = function(cg)
+    cg.s:exec(function(wait_mode)
+        -- Fill up the limbo.
+        local sync_f = _G.fiber.create(function()
+            box.space.s:replace{0}
+        end)
+        sync_f:set_joinable(true)
+        t.assert_equals(box.info.synchro.queue.len, 1)
+        t.assert_ge(box.info.synchro.queue.size,
+                    box.cfg.replication_synchro_queue_max_size)
+
+        if wait_mode == 'none' then
+            local msg = 'The synchronous transaction queue is full'
+            t.assert_error_msg_content_equals(msg, function()
+                box.atomic({wait = 'none'}, function()
+                    box.space.s:replace{1}
+                end)
+            end)
+            return
+        end
+
+        local atomic_f = _G.fiber.create(function()
+            box.atomic({wait = wait_mode}, function()
+                box.space.s:replace{1}
+            end)
+        end)
+        atomic_f:set_joinable(true)
+        t.assert_equals(box.info.synchro.queue.len, 1)
+
+        box.cfg{replication_synchro_quorum = 1}
+        t.assert(sync_f:join())
+        box.cfg{replication_synchro_quorum = 2}
+        t.assert_equals(box.space.s:get{0}, {0})
+        t.assert_equals(box.info.synchro.queue.len, 1)
```

Comment:
Why is it guaranteed 1? You had a moment after `box.cfg{replication_synchro_quorum = 1}` when there was a yield. Technically the second transaction during that yield could reach WAL, get written, produce a `CONFIRM`, and get committed. All before you restore `box.cfg{replication_synchro_quorum = 2}`.

Reply:
I see your concern, but the fiber responsible for the second transaction will have to yield after submitting the journal request and AFAIC it is guaranteed that it will be scheduled later than the fiber running the test code. I have also ran many test instances concurrently with a large number of test-run jobs, and it doesn't fail.

Reply:
Hm. So the test entirely depends on the fiber scheduling order. That looks quite unstable, but I don't mind fixing it later, when it starts flakying.